src/main/piece/tetris_piece.cpp

```diff
@@ -195,10 +182,43 @@ void TetrisPiece::shift( int direction_unit )
 
 void TetrisPiece::set_block_locations( std::vector<std::unique_ptr<Point>> param_block_locations )
 {
-  block_locations = move( param_block_locations );
+  current_block_locations = move( param_block_locations );
+}
+
+void TetrisPiece::set_next_block_locations( vector<unique_ptr<Point>> param_block_locations )
+{
+  next_block_locations = move( param_block_locations );
 } 
 
 void TetrisPiece::add_block_location( std::unique_ptr<Point> point )
 {
-  block_locations.push_back( move( point ) );
+  current_block_locations.push_back( move( point ) );
+}
+
+vector<unique_ptr<Point>> TetrisPiece::rotate_block_locations()
+{
+  int minimum_x = 0;
+  vector<unique_ptr<Point>> rotated_block_locations;
+  for( auto& point : current_block_locations )
+  {
+    int new_x = -1 * point -> get_y();
+    int new_y = point -> get_x();
+    unique_ptr<Point> new_point = make_unique<Point>( new_x, new_y );
+
+    rotated_block_locations.push_back( move( new_point ) );
+    if( new_x < minimum_x )
+    {
+      minimum_x = new_x;
+    }
+  }
+
+  if( minimum_x < 0 )
+  {
+    for( auto& point : rotated_block_locations )
+    {
+      point -> set_x( point -> get_x() + abs( minimum_x ) );
```

Comment:
The piece locations habe to be translated back into the first quadrant, so the piece doesn't shift during rotation.